internal/runtime/internal/controller/component_references.go

```diff
@@ -63,6 +64,24 @@ func ComponentReferences(cn dag.Node, g *dag.Graph, l log.Logger, scope *vm.Scop
 			refs = append(refs, ref)
 		} else if scope.IsStdlibDeprecated(t[0].Name) {
 			level.Warn(l).Log("msg", "this stdlib function is deprecated; please refer to the documentation for updated usage and alternatives", "function", t[0].Name)
+		} else {
+			var tokenNames []string
+			for _, token := range t {
+				tokenNames = append(tokenNames, token.Name)
+			}
+
+			funcName := strings.Join(tokenNames, ".")
+			if scope.IsStdlibExperimental(funcName) {
+				if err := featuregate.CheckAllowed(featuregate.StabilityExperimental, minStability, funcName); err != nil {
```

Comment:
So we have `Scope` that knows whether funcName is experimental.
And we have this function - `ComponentReferences` knowing the `minStability`. 

What if we had `Scope` knowing both pieces of information - the `minStability` and the funcNames that are experimental and it would return error or diags right away to here?

If that has some implementation challenges, I'm open to stay at what it is now, not to further delay this. But it seems fragmented and a bit complicated now.

Reply:
I think that makes sense but we would need to bring the featuregate pkg outside of the internal pkg so that the syntax pkg has access to it + some changes in the syntax pkg to adapt it. It's not that difficult but it will bloat this PR. I would suggest to do this in a follow up PR. I'm already planning to refactor the stdlib because the names in the code are not great right now

Reply:
ok! makes perfect sense, happy to do these later